File: javascript/HouseRobber.js
// https://leetcode.com/problems/house-robber/

/**
 * @param {number[]} nums
 * @return {number}
 */
function rob(nums) {
  'use strict' // Allow block scoping

  // Edge cases
  if (nums.length === 0) return 0;
  if (nums.length === 1) return nums[0];
  if (nums.length === 2) return Math.max(nums[0], nums[1]);

  // Store the results
  let max1 = 0
  let max2 = 0

  for (let money of nums) {
    const haul = max1 + money   // max(i - 1) + L[i]
    max1 = Math.max(max2, max1) // max(max(i - 2) + L[i - 1], max(i - 1))
    max2 = haul                 // Assign the haul
  }

  return Math.max(max2, max1)
}

File: c++/HouseRobber.cpp
// https://leetcode.com/problems/house-robber/

/*
  You are a professional robber planning to rob houses along a street. 
  Each house has a certain amount of money stashed, the only constraint 
  stopping you from robbing each of them is that adjacent houses have 
  security system connected and it will automatically contact the police 
  if two adjacent houses were broken into on the same night.

  Given a list of non-negative integers representing the amount of 
  money of each house, determine the maximum amount of money you can 
  rob tonight without alerting the police.
*/

// f(n) = max(rob(n - 1), rob(n, n - 2))

#include <vector>

class Solution {
public:
  int rob(std::vector<int>& nums) {
    unsigned robberies[3] = { 0, 0, 0 };
    for (int house : nums) {
      robberies[0] = robberies[1];
      robberies[1] = robberies[2];
      robberies[2] = std::max(robberies[1], house + robberies[0]);
    }
    return robberies[2];
  }
};
